### ISSUE TITLE
Regression in 2.14: batched inference produces different output than single inference for vision-language models

### ISSUE THREAD

**Issue by ydshieh** [pytorch/pytorch#196886]
# PyTorch Issue: Batched inference produces different output than single inference on torch 2.14 (regression from 2.13)

Hi, I'm @ydshieh from the HuggingFace Transformers team. We were running a CI with the released torch 2.14 before switching to it, and found a behavior worth reporting here.

## Summary

We are **not** reporting the usual output value drift between torch versions (where golden values shift slightly — that is expected and acceptable). This is specifically about the **`single == batch` property being broken in torch 2.14**.

When running the **same input** through a vision-language model in two ways:
1. **Single** — one example, no padding
2. **Batch** — same example as the first item in a batch (padded to match a second, longer example)

...these two should produce identical output. This property holds on **torch 2.13** but is broken on **torch 2.14**.

We observed this with at least two independently-developed model families:
- `ibm-granite/granite-vision-4.1-4b` (Granite4Vision)
- `llava-hf/LLaVA-NeXT-Video-7B-hf` (LLaVA-NeXT-Video)

**Remark**: despite `single != batch` on torch 2.14, both outputs still look semantically reasonable on their own — this is not a case of garbled or nonsensical text. We want to flag it to the PyTorch team to double-check whether this property being broken is an expected consequence of new optimizations or an unintended side effect.

---

## Environment

```
torch==2.14.0+cu130
transformers @ git+https://github.com/huggingface/transformers@ff4bb5ee67be830a9bd45426ca5626825c3f222a
CUDA 13.0 (driver), A10G GPU (compute capability 8.6)
```

### Setup

```bash
git clone https://github.com/huggingface/transformers.git
cd transformers
git checkout ff4bb5ee67be830a9bd45426ca5626825c3f222a
pip install -e ".[dev]"

# Install torch 2.14
pip install --no-cache-dir torch torchvision --index-url https://download.pytorch.org/whl/cu130

# Verify
python -c "import torch; print(torch.__version__)"  # should print 2.14.0+cu130
```

---

## Reproducer 1 — `ibm-granite/granite-vision-4.1-4b`

```python
import torch
from PIL import Image
from transformers import AutoProcessor, Granite4VisionForConditionalGeneration
from huggingface_hub import hf_hub_download

model_id = "ibm-granite/granite-vision-4.1-4b"
processor = AutoProcessor.from_pretrained(model_id)
model = Granite4VisionForConditionalGeneration.from_pretrained(
    model_id,
    torch_dtype=torch.bfloat16,
    attn_implementation={"qformer_config": "eager"},
).to("cuda")

# Load two test images
from transformers.image_utils import load_image
image1 = load_image("https://huggingface.co/datasets/hf-internal-testing/fixtures-coco/resolve/main/val2017/000000039769.jpg")
image2 = load_image("https://huggingface.co/datasets/hf-internal-testing/fixtures-coco/resolve/main/val2017/000000001000.jpg")

def make_prompt(processor, question):
    messages = [{"role": "user", "content": [{"type": "image"}, {"type": "text", "text": question}]}]
    return processor.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)

prompt = make_prompt(processor, "What do you see in this image?")

# --- Single inference (image1 only) ---
inputs_single = processor(text=prompt, images=image1, return_tensors="pt").to("cuda")
output_single = model.generate(**inputs_single, max_new_tokens=30, do_sample=False)
decoded_single = processor.decode(
    output_single[0, inputs_single["input_ids"].shape[1]:], skip_special_tokens=True
)

# --- Batch inference (image1 + image2, padded) ---
inputs_batch = processor(
    text=[prompt, prompt],
    images=[image1, image2],
    return_tensors="pt",
    padding=True,
).to("cuda")
output_batch = model.generate(**inputs_batch, max_new_tokens=30, do_sample=False)
decoded_batch = processor.decode(
    output_batch[0, inputs_batch["input_ids"].shape[1]:], skip_special_tokens=True
)

print("Single:", decoded_single)
print("Batch[0]:", decoded_batch)
print("Match:", decoded_single == decoded_batch)
```

### Expected output (torch 2.13)

Both `single` and `batch[0]` produce the **same** text:

```
Single:  The image depicts two cats resting on a pink blanket. They are lying in a relaxed, sprawled position, with one cat appearing to be in a
Batch[0]: The image depicts two cats resting on a pink blanket. They are lying in a relaxed, sprawled position, with one cat appearing to be in a
Match: True
```

### Actual output (torch 2.14)

```
Single:  The image depicts two cats resting on a bright pink blanket spread over a piece of furniture, likely a couch. The cat on the left is lying on
Batch[0]: I see two cats lying on a pink blanket. One cat is on the left side, and the other is on the right side. There are two
Match: False
```

The single and batch outputs are not just numerically different — they have completely different sentence structure, suggesting the attention values in the padded batch are substantially wrong.

### Cross-version results (Reproducer 1)

Tested on A10G (cu130), all other deps fixed:

| torch | Match |
|---|---|
| 2.11.0 | ✅ True |
| 2.12.1 | ✅ True |
| 2.13.0 | ✅ True |
| 2.14.0 | ❌ False |

The regression is clean — only torch 2.14 breaks this.

---

## Reproducer 2 — `llava-hf/LLaVA-NeXT-Video-7B-hf`

```python
import torch
import numpy as np
from PIL import Image
from transformers import AutoProcessor, LlavaNextVideoForConditionalGeneration, BitsAndBytesConfig
from huggingface_hub import hf_hub_download

processor = AutoProcessor.from_pretrained("llava-hf/LLaVA-NeXT-Video-7B-hf")
model = LlavaNextVideoForConditionalGeneration.from_pretrained(
    "llava-hf/LLaVA-NeXT-Video-7B-hf",
    quantization_config=BitsAndBytesConfig(load_in_4bit=True),
)

image_file = hf_hub_download(repo_id="raushan-testing-hf/images_test", filename="llava_v1_5_radar.jpg", repo_type="dataset")
video_file = hf_hub_download(repo_id="raushan-testing-hf/videos-test", filename="video_demo.npy", repo_type="dataset")
image = Image.open(image_file)
video = np.load(video_file)

prompt_video = "USER: <video>\nWhy is this video funny? ASSISTANT:"
prompt_image = "USER: <image>\nWhat is shown in this image? ASSISTANT:"

# --- Batch: video prompt + image prompt together (padding required) ---
inputs_batched = processor(
    text=[prompt_video, prompt_image],
    images=[image],
    videos=[video],
    return_tensors="pt",
    padding=True,
).to("cuda")

# --- Single: video prompt alone (no padding) ---
inputs_single = processor(
    text=prompt_video,
    videos=[video],
    return_tensors="pt",
).to("cuda")

output_batched = model.generate(**inputs_batched, do_sample=False, max_new_tokens=50)
output_single = model.generate(**inputs_single, do_sample=False, max_new_tokens=50)

decoded_batched = processor.decode(output_batched[0], skip_special_tokens=True)
decoded_single = processor.decode(output_single[0], skip_special_tokens=True)

print("Batched[0]:", decoded_batched)
print("Single:    ", decoded_single)
print("Match:", decoded_batched == decoded_single)
```

### Expected output (torch 2.13)

```
Batched[0]: USER: Why is this video funny? ASSISTANT: The humor in this video comes from the unexpected and somewhat comical situation of a young child reading a book while wearing a pair of glasses that are too large for her...
Single:     USER: Why is this video funny? ASSISTANT: The humor in this video comes from the unexpected and somewhat comical situation of a young child reading a book while wearing a pair of glasses that are too large for her...
Match: True
```

### Actual output (torch 2.14)

```
Batched[0]: USER: Why is this video funny? ASSISTANT: The humor in this video comes from the unexpected and somewhat comical situation of a young child reading a book while another child is attempting to read the same book. The child who is reading the book seems to be struggling...
Single:     USER: Why is this video funny? ASSISTANT: The humor in this video comes from the unexpected and somewhat comical situation of a young child reading a book while wearing a pair of glasses that are too large for her. The glasses are so large that they cover her eyes...
Match: False
```

Again: the batched item's output is different from the single output. Both are semantically reasonable on their own — this is not garbled text. But the `single == batch` property no longer holds.

### Cross-version results (Reproducer 2)

Tested on A10G (cu130), all other deps fixed:

| torch | Match | Notes |
|---|---|---|
| 2.11.0 | ✅ True | |
| 2.12.1 | ❌ False | Minor pronoun difference only ("her" → "them", same scenario — likely 4-bit quantization non-determinism rather than a torch regression) |
| 2.13.0 | ✅ True | |
| 2.14.0 | ❌ False | Completely different scenario described |

The 2.12.1 flicker appears to be quantization noise (the outputs describe the same scenario with slightly different pronouns). The 2.14 failure is qualitatively different — the batch output describes an entirely different situation.

---

## Additional evidence — audio output model (`Qwen/Qwen3-Omni-30B-A3B-Instruct`)

We see the same property break in an audio-output model. When running two **identical prompts** in a batch, the two output waveforms should be identical. On torch 2.13 they are **bit-for-bit identical** (max abs diff = 0.0). On torch 2.14, 125 out of 16725 audio samples differ with max abs diff ~3.17e-3.

This is a stricter version of the same issue: not single-vs-batch, but row[0]-vs-row[1] within a batch of duplicate inputs. The divergence in 2.14 is not just a looser tolerance — the property that was perfectly exact in 2.13 is now broken.

---

## Questions for the PyTorch team

1. **Were new optimizations added in torch 2.14 that could affect this behavior?** For example, changes to SDPA dispatch, fused attention kernels, or attention mask handling for padded batches that might alter numerical results compared to 2.13?

2. **Are there any flags or environment variables to restore the 2.13 behavior?** For example, disabling a specific kernel path (`TORCH_SDPA_BACKEND`, `torch.backends.cuda.enable_flash_sdp(False)`, or similar) so that padded batch inference matches single inference again?

We have a large test suite that relies on `single == batch` as a correctness invariant. A workaround flag would let us unblock CI while the root cause is investigated.

---

## References

- Transformers verification run on torch 2.14: https://github.com/huggingface/transformers/actions/runs/34686882433
- Torch 2.13 baseline CI (tests pass): https://github.com/huggingface/transformers/actions/runs/34555198499
- `granite4_vision` test: `tests/models/granite4_vision/test_modeling_granite4_vision.py::Granite4VisionIntegrationTest::test_small_model_integration_test_batch_matches_single`
- `llava_next_video` test: `tests/models/llava_next_video/test_modeling_llava_next_video.py::LlavaNextVideoForConditionalGenerationIntegrationTest::test_small_model_integration_test_batch_matches_single`

cc @ezyang @gchanan @kadeng @msaroufim

**Comment by drisspg** [comment]
Seems like the main change was this pr:  https://github.com/pytorch/pytorch/pull/187091

   | PyTorch | cuDNN |
   |---|---|
   | 2.13.0+cu130 | 9.20.0.48 |
   | 2.14.0+cu130 | 9.24.0.43 |


Since this a10g I don think sdpa is involved here at all and this really about the convolutions. At least on my machine with gb300 it does not show different sampling though at single vs batched. For testing can you try setting `   torch.backends.cudnn.enabled = False` 

This in general is not a property we guarantee though as well.

**Comment by gderossi** [comment]
I've confirmed that the cuDNN update was the trigger for this behavior, though ultimately this is caused by batch variance coming from GEMMs (cuBLAS does not guarantee batch invariance). The conv heuristic in cuDNN 9.24 ranks algorithms differently than 9.20. For the particular case in reproducer 1, it looks like 9.24 prioritizes an algorithm which does not require a workspace. This algorithm is batch-invariant but returns a slightly different output than the algorithm selected by 9.20, and the following batch-variant GEMMs then cause the divergence.

EDIT: Are there expectations/hopes for batch-invariant GEMMs? It's something we've discussed with cuBLAS before but it has never been prioritized.

Here's a diagram from my agent explaining the behavior for reproducer 1:

<img width="2235" height="1410" alt="Image" src="https://github.com/user-attachments/assets/eb3ee8cb-7fc4-4661-9570-96654876c2cd" />

**Comment by drisspg** [comment]
I dont think that this is high pri and not release 2.14.1 blocking; while in general batch invariance is a nice property and something that if all impls are equal would should choose this path it is not something that we currently gurantee. 

I am going to leave the issue open; but we might just want to convert this into an RFC cc @eqy
